domain/secretbackend/state/state.go

```diff
@@ -720,6 +795,279 @@ func (s *State) getModelCloudAndCredential(
+	AND secret_revision_uuid = $SecretBackendReference.secret_revision_uuid`, refCount)
+	if err != nil {
+		return SecretBackendReference{}, errors.Trace(err)
+	}
+
+	err = tx.Query(ctx, stmt, refCount).Get(&refCount)
+	if errors.Is(err, sql.ErrNoRows) {
+		return SecretBackendReference{}, fmt.Errorf(
+			"%w: model %q secret revision %q",
+			secretbackenderrors.RefCountNotFound, modelID, revisionID,
+		)
+	}
+	if err != nil {
+		return SecretBackendReference{}, fmt.Errorf(
+			"cannot get secret backend reference for model %q secret revision %q: %w",
+			modelID, revisionID, err,
+		)
+	}
+	return refCount, nil
+}
+
+// UpdateSecretBackendReference updates the reference to the secret backend for the given secret revision, returning an error
+// satisfying [secretbackenderrors.RefCountNotFound] if no existing refcount was found.
+// It returns a rollback function which can be used to revert the changes.
+func (s *State) UpdateSecretBackendReference(
+	ctx context.Context, valueRef *secrets.ValueRef, modelID coremodel.UUID, revisionID uuid.UUID,
+) (func() error, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var existing SecretBackendReference
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		existing, err = s.getSecretBackendReferenceBackendID(ctx, tx, modelID, revisionID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		backendID, err := s.getSecretBackendID(ctx, tx, valueRef)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if err := s.removeSecretBackendReferenceForRevisions(ctx, tx, revisionID); err != nil {
+			return errors.Trace(err)
+		}
+		if err := s.addSecretBackendReference(ctx, tx, backendID, modelID, revisionID); err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return func() error {
+		err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+			if err := s.removeSecretBackendReferenceForRevisions(ctx, tx, revisionID); err != nil {
+				return errors.Trace(err)
+			}
+			err := s.addSecretBackendReference(ctx, tx, existing.BackendID, modelID, revisionID)
+			return errors.Trace(err)
+		})
+		return errors.Trace(err)
+	}, nil
+}
+
+type secretRevisionIDs []string
+
+// RemoveSecretBackendReference removes the reference to the secret backend for the given secret revision.
+func (s *State) RemoveSecretBackendReference(ctx context.Context, revisionIDs ...uuid.UUID) error {
+	if len(revisionIDs) == 0 {
+		return nil
+	}
+
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := s.removeSecretBackendReferenceForRevisions(ctx, tx, revisionIDs...)
+		return errors.Trace(err)
+	})
+	return errors.Trace(err)
+}
+
+func (s *State) removeSecretBackendReferenceForRevisions(ctx context.Context, tx *sqlair.TX, revisionIDs ...uuid.UUID) error {
```

Comment:
Not sure this needs to be its own function.
Even if it were, it has essentially the same signature as the public function about but uses a different name.

Reply:
This is used for two exported methods.